radix-engine/src/system/system.rs

```diff
@@ -1430,63 +1447,67 @@ where
                         }
 
                         // Anything frame owned does not have a global address
-                        Visibility::FrameOwned => return None,
+                        Visibility::FrameOwned => return Ok(ReceiverType::OnHeapObject),
 
                         // If borrowed or actor then we just use the current actor's global address
                         // e.g. if the parent to the node is frame owned then the current actor's global
                         // address would be None
                         Visibility::Borrowed | Visibility::Actor => {
-                            return self
-                                .api
-                                .kernel_get_system_state()
-                                .current_actor
-                                .global_address();
+                            match self.api.kernel_get_system_state().current_actor {
+                                Actor::Method(MethodActor { receiver_type, .. }) => {
+                                    let method_type = match receiver_type {
+                                        ReceiverType::DirectAccess => ReceiverType::DirectAccess,
```

Comment:
I think the state transition is wrong here. If the visibility is obtained by reading a heap object (such as kv store), the new receiver type shouldn't be `DirectAccess`.

The new receiver type depends on whether the reference comes from, which requires another refactor to the reference tracking.

Luckily, direct access is only exposed to `Vault` atm, so suggest adding some comments and moving on.